src/ExponentialBackoff.php

```diff
@@ -0,0 +1,158 @@
+<?php
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud;
+
+/**
+ * Exponential backoff implementation.
+ */
+class ExponentialBackoff
+{
+    const MAX_DELAY_MICROSECONDS = 60000000;
+
+    /**
+     * @var int
+     */
+    private $retries;
+
+    /**
+     * @var callable
+     */
+    private $delayFunction;
+
+    /**
+     * @var array
+     */
+    private $httpRetryCodes = [
+        500,
+        502,
+        503,
+        504
+    ];
+
+    /**
+     * @var array
+     */
+    private $httpRetryMessages = [
+        'rateLimitExceeded',
+        'userRateLimitExceeded'
+    ];
+
+    /**
+     * @param int $retries Number of retries for a failed request.
+     */
+    public function __construct($retries = null)
+    {
+        $this->retries = $retries !== null ? (int) $retries : 3;
+        // @todo revisit this approach
+        // @codeCoverageIgnoreStart
+        $this->delayFunction = function ($delay) {
+            usleep($delay);
+        };
+        // @codeCoverageIgnoreEnd
+    }
+
+    /**
+     * Executes the retry process.
+     *
+     * @param callable $function
+     * @param array $arguments
+     * @return mixed
+     * @throws \Exception
```

Comment:
I think we should return the last one. If the exceptions are 503, 503, 404, you won't see there's 404.
Also you should check shouldRetry everytime (see other note).